src/features/island/hud/components/AddSFL.tsx

```diff
@@ -0,0 +1,179 @@
+import React, { ChangeEvent, useContext, useEffect, useState } from "react";
+import classNames from "classnames";
+import { CloseButtonPanel } from "features/game/components/CloseablePanel";
+import { Modal } from "react-bootstrap";
+import token from "assets/icons/token_2.png";
+
+import matic from "assets/icons/polygon-token.png";
+import { Button } from "components/ui/Button";
+import { wallet } from "lib/blockchain/wallet";
+import { fromWei, toBN, toWei } from "web3-utils";
+import Decimal from "decimal.js-light";
+import { setPrecision } from "lib/utils/formatNumber";
+import { Context } from "features/game/GameProvider";
+
+interface Props {
+  isOpen: boolean;
+  onClose: () => void;
+}
+
+const VALID_NUMBER = new RegExp(/^\d*\.?\d*$/);
+const INPUT_MAX_CHAR = 10;
+
+export const AddSFL: React.FC<Props> = ({ isOpen, onClose }) => {
+  const { gameService } = useContext(Context);
+  const [maticBalance, setMaticBalance] = useState(0);
+  const [maticAmount, setMaticAmount] = useState(0);
+  const [SFLAmount, setSFLAmount] = useState(0);
+  const [maticInputError, setMaticInputError] = useState<string | null>(null);
+
+  const amountOutMin = SFLAmount * 0.99;
+
+  useEffect(() => {
+    const fetchMaticBalance = async () => {
+      const balance = await wallet.getMaticBalance();
+
+      setMaticBalance(balance);
+    };
+
+    fetchMaticBalance();
+  }, []);
+
+  const getSFLForMaticAmount = async (amount: number) => {
+    try {
+      const sfl = await wallet.getSFLForMatic(toWei(amount.toString()));
+
+      if (sfl) {
+        setMaticInputError(null);
+      }
+
+      setSFLAmount(sfl);
+    } catch (error: any) {
+      console.error(error.message);
+    }
+  };
+
+  const handleMaticAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
+    // Strip the leading zero from numbers
+    if (/^0+(?!\.)/.test(e.target.value) && e.target.value.length > 1) {
+      e.target.value = e.target.value.replace(/^0/, "");
+    }
+
+    if (VALID_NUMBER.test(e.target.value)) {
+      const input = Number(e.target.value.slice(0, INPUT_MAX_CHAR));
+
+      setMaticAmount(input);
+
+      if (input === 0) {
+        setSFLAmount(0);
+      } else {
+        getSFLForMaticAmount(input);
+      }
+    }
+  };
+
+  const handleAddSFL = () => {
+    gameService.send({
+      type: "BUY_SFL",
+      maticAmount: toWei(maticAmount.toString()),
+      amountOutMin: toWei(amountOutMin.toString()),
+    });
+    onClose();
+  };
+
+  const maticBalString = fromWei(toBN(maticBalance));
```

Comment:
I am getting an error once the game has loaded.

It appears that my `maticBalance` is creating some issues.

<img width="359" alt="Screen Shot 2022-12-28 at 2 39 20 pm" src="https://user-images.githubusercontent.com/11745561/209753565-af04d595-7883-4109-98eb-6e1c33fd0671.png">
<img width="795" alt="Screen Shot 2022-12-28 at 2 39 28 pm" src="https://user-images.githubusercontent.com/11745561/209753568-89f1dd6c-d6cc-4fc1-8feb-3d3045397fff.png">
<img width="328" alt="Screen Shot 2022-12-28 at 2 39 59 pm" src="https://user-images.githubusercontent.com/11745561/209753571-09767f04-0312-48ad-83bc-90f0cfb28a8a.png">